# Data Cleaning - Cricket Test Matches

**Data source:** https://www.espncricinfo.com/records/highest-career-batting-average-282910

In [2]:
# import libraries
import pandas as pd

In [3]:
# load data
df = pd.read_excel('Cricket_Test_Matches_Data.xlsx')

# look at data 
df.head()

Player       Span  Mat  Inns  NO  Runs    HS    Ave     BF  \
0  DG Bradman (AUS)  1928-1948   52    80  10  6996   334  99.94  9800+   
1    HC Brook (ENG)  2022-2023   12    20   1  1181   186  62.15   1287   
2    AC Voges (AUS)  2015-2016   20    31   7  1485  269*  61.87   2667   
3   RG Pollock (SA)  1963-1970   23    41   4  2256   274  60.97  1707+   
4   GA Headley (WI)  1930-1954   22    40   4  2190  270*  60.83   416+   

      SR  100  50  0   4s  6s  
0  58.60   29  13  7  626   6  
1  91.76    4   7  1  141  23  
2  55.68    5   4  2  186   5  
3  54.48    7  11  1  246  11  
4  56.00   10   5  2  104   1

We may notice a couple of inaccuracies in the data. Like `270*` or `269*`

**Rename columns**

In [4]:
# rename columns to make them comprehensive 
df = df.rename(columns = {'NO':'not_outs', 'Mat':'matches', 'HS':'highest_inns_score', 'BF':'balls_faced', 'SR':'batting_strike_rate'})

In [5]:
df.head(1)

Player       Span  matches  Inns  not_outs  Runs  \
0  DG Bradman (AUS)  1928-1948       52    80        10  6996   

  highest_inns_score    Ave balls_faced  batting_strike_rate  100  50  0   4s  \
0                334  99.94       9800+                 58.6   29  13  7  626   

   6s  
0   6

In [6]:
# rename columns to make them in consistent format
df = df.rename(columns = {'Span':'span', 'Player':'player', 'Runs':'runs', 'Ave':'ave', 'SR':'batting_strike_rate'})

In [8]:
df.head(1)

player       span  matches  Inns  not_outs  runs  \
0  DG Bradman (AUS)  1928-1948       52    80        10  6996   

  highest_inns_score    ave balls_faced  batting_strike_rate  100  50  0   4s  \
0                334  99.94       9800+                 58.6   29  13  7  626   

   6s  
0   6

**Deal with `null` values**

In [9]:
df.isnull().sum()

player                 0
span                   0
matches                0
Inns                   0
not_outs               0
runs                   0
highest_inns_score     0
ave                    0
balls_faced            3
batting_strike_rate    3
100                    0
50                     0
0                      0
4s                     0
6s                     0
dtype: int64

In [10]:
df[df['balls_faced'].isna()]

player       span  matches  Inns  not_outs  runs  \
8         ED Weekes (WI)  1948-1958       48    81         5  4455   
16       CL Walcott (WI)  1948-1960       44    74         7  3798   
57  Hon.FS Jackson (ENG)  1893-1905       20    33         4  1415   

   highest_inns_score    ave balls_faced  batting_strike_rate  100  50  0  \
8                 207  58.61         NaN                  NaN   15  19  6   
16                220  56.68         NaN                  NaN   15  14  1   
57               144*  48.79         NaN                  NaN    5   6  3   

     4s  6s  
8   258   2  
16  107  11  
57   51   0

In [11]:
df['balls_faced'] = df['balls_faced'].fillna(0)
df['batting_strike_rate'] = df['batting_strike_rate'].fillna(0)

In [12]:
df[df['player'] == 'ED Weekes (WI)']

player       span  matches  Inns  not_outs  runs  \
8  ED Weekes (WI)  1948-1958       48    81         5  4455   

  highest_inns_score    ave balls_faced  batting_strike_rate  100  50  0   4s  \
8                207  58.61           0                  0.0   15  19  6  258   

   6s  
8   2

**Deal with duplicates**

In [13]:
df[df['player'].duplicated() == 1]

player       span  matches  Inns  not_outs   runs  \
10    WR Hammond (ENG)  1927-1947       85   140        16   7249   
18  JH Kallis (ICC/SA)  1995-2013      166   280        40  13289   
23       CA Davis (WI)  1968-1973       15    29         5   1301   

   highest_inns_score    ave balls_faced  batting_strike_rate  100  50   0  \
10               336*  58.45       7491+                38.07   22  24   4   
18                224  55.37       28903                45.97   45  58  16   
23                183  54.20        665+                35.48    4   4   1   

      4s  6s  
10   419  27  
18  1488  97  
23    65   1

In [14]:
# look at duplicated rows
df[df['player'].isin(['WR Hammond (ENG)', 'JH Kallis (ICC/SA)', 'CA Davis (WI)'])]

player       span  matches  Inns  not_outs   runs  \
9     WR Hammond (ENG)  1927-1947       85   140        16   7249   
10    WR Hammond (ENG)  1927-1947       85   140        16   7249   
13  JH Kallis (ICC/SA)  1995-2013      166   280        40  13289   
18  JH Kallis (ICC/SA)  1995-2013      166   280        40  13289   
21       CA Davis (WI)  1968-1973       15    29         5   1301   
23       CA Davis (WI)  1968-1973       15    29         5   1301   

   highest_inns_score    ave balls_faced  batting_strike_rate  100  50   0  \
9                336*  58.45       7491+                38.07   22  24   4   
10               336*  58.45       7491+                38.07   22  24   4   
13                224  55.37       28903                45.97   45  58  16   
18                224  55.37       28903                45.97   45  58  16   
21                183  54.20        665+                35.48    4   4   1   
23                183  54.20        665+                35.48    4   4   1   

      4s  6s  
9    419  27  
10   419  27  
13  1488  97  
18  1488  97  
21    65   1  
23    65   1

In [16]:
# drop duplicates
df = df.drop_duplicates()

In [17]:
df[df['player'].duplicated() == 1]

Empty DataFrame
Columns: [player, span, matches, Inns, not_outs, runs, highest_inns_score, ave, balls_faced, batting_strike_rate, 100, 50, 0, 4s, 6s]
Index: []

No output, therefore there are no duplicated rows in our data 

**Split up the `span` column into start_date and end_date**

In [18]:
df['span'].str.split(pat = '-')

0     [1928, 1948]
1     [2022, 2023]
2     [2015, 2016]
3     [1963, 1970]
4     [1930, 1954]
          ...     
60    [1965, 1981]
61    [2002, 2014]
62    [1924, 1934]
63    [1930, 1938]
64    [1928, 1934]
Name: span, Length: 62, dtype: object

In [19]:
df['start_date'] = df['span'].str.split(pat = '-').str[0]
df['end_date'] = df['span'].str.split(pat = '-').str[1]

In [20]:
df.head()

player       span  matches  Inns  not_outs  runs  \
0  DG Bradman (AUS)  1928-1948       52    80        10  6996   
1    HC Brook (ENG)  2022-2023       12    20         1  1181   
2    AC Voges (AUS)  2015-2016       20    31         7  1485   
3   RG Pollock (SA)  1963-1970       23    41         4  2256   
4   GA Headley (WI)  1930-1954       22    40         4  2190   

  highest_inns_score    ave balls_faced  batting_strike_rate  100  50  0   4s  \
0                334  99.94       9800+                58.60   29  13  7  626   
1                186  62.15        1287                91.76    4   7  1  141   
2               269*  61.87        2667                55.68    5   4  2  186   
3                274  60.97       1707+                54.48    7  11  1  246   
4               270*  60.83        416+                56.00   10   5  2  104   

   6s start_date end_date  
0   6       1928     1948  
1  23       2022     2023  
2   5       2015     2016  
3  11       1963     1970  
4   1       1930     1954

In [21]:
# drop span column
df = df.drop(['span'], axis = 1)

In [22]:
df.head(1)

player  matches  Inns  not_outs  runs highest_inns_score    ave  \
0  DG Bradman (AUS)       52    80        10  6996                334  99.94   

  balls_faced  batting_strike_rate  100  50  0   4s  6s start_date end_date  
0       9800+                 58.6   29  13  7  626   6       1928     1948

**Split up the Country from the player**

In [23]:
df['country'] = df['player'].str.split(pat = '(').str[1]

In [24]:
df['country'] = df['country'].str.split(pat = ')').str[0]

In [25]:
df['country'].head()

0    AUS
1    ENG
2    AUS
3     SA
4     WI
Name: country, dtype: object

It works, but we mat notice that country is still in the `player` columns:

In [26]:
df['player'].head()

0    DG Bradman (AUS)
1      HC Brook (ENG)
2      AC Voges (AUS)
3     RG Pollock (SA)
4     GA Headley (WI)
Name: player, dtype: object

In [27]:
df['player'] = df['player'].str.split(pat = '(').str[0]

In [28]:
df.player.head()

0    DG Bradman 
1      HC Brook 
2      AC Voges 
3    RG Pollock 
4    GA Headley 
Name: player, dtype: object

**Change data types**

In [29]:
df.dtypes

player                  object
matches                  int64
Inns                     int64
not_outs                 int64
runs                     int64
highest_inns_score      object
ave                    float64
balls_faced             object
batting_strike_rate    float64
100                      int64
50                       int64
0                        int64
4s                       int64
6s                       int64
start_date              object
end_date                object
country                 object
dtype: object

In [35]:
df['highest_inns_score'].str.split(pat = '*').str[0].head()

# df['highest_inns_score'].str.split(pat = '*').str[1] 

# str[1] is the '*' symbol, so I will replace columns with only str[0] values

0    334
1    186
2    269
3    274
4    270
Name: highest_inns_score, dtype: object

In [34]:
# Clean highest_inns_score column from '*'
df['highest_inns_score'] = df['highest_inns_score'].str.split(pat = '*').str[0] 

In [40]:
# Change highest_inns_score data type from object to int
df['highest_inns_score'] = df['highest_inns_score'].astype('int')

# Change start_date and end_date data types from object to int
df = df.astype({'start_date':'int', 'end_date':'int'})

In [44]:
# Clean balls_faced column from '+'
df['balls_faced'] = df['balls_faced'].str.split(pat = '+').str[0]

In [45]:
df[df['balls_faced'].isna()]

player  matches  Inns  not_outs  runs  highest_inns_score    ave  \
8        ED Weekes        48    81         5  4455                 207  58.61   
16      CL Walcott        44    74         7  3798                 220  56.68   
57  Hon.FS Jackson        20    33         4  1415                 144  48.79   

   balls_faced  batting_strike_rate  100  50  0   4s  6s  start_date  \
8          NaN                  0.0   15  19  6  258   2        1948   
16         NaN                  0.0   15  14  1  107  11        1948   
57         NaN                  0.0    5   6  3   51   0        1893   

    end_date country  
8       1958      WI  
16      1960      WI  
57      1905     ENG

We see that `balls_faced` column has **NaN** values, so we have to deal with them to convert data type

In [46]:
df['balls_faced'] = df['balls_faced'].fillna(0)

In [47]:
df.balls_faced.head()

0    9800
1    1287
2    2667
3    1707
4     416
Name: balls_faced, dtype: object

In [48]:
# Сonvert type to int
df['balls_faced'] = df['balls_faced'].astype('int')

In [49]:
df.dtypes

player                  object
matches                  int64
Inns                     int64
not_outs                 int64
runs                     int64
highest_inns_score       int32
ave                    float64
balls_faced              int32
batting_strike_rate    float64
100                      int64
50                       int64
0                        int64
4s                       int64
6s                       int64
start_date               int32
end_date                 int32
country                 object
dtype: object

**Create a new column with career length**

In [54]:
df['career_length'] = df['end_date'] - df['start_date']

In [55]:
df.head()

player  matches  Inns  not_outs  runs  highest_inns_score    ave  \
0  DG Bradman        52    80        10  6996                 334  99.94   
1    HC Brook        12    20         1  1181                 186  62.15   
2    AC Voges        20    31         7  1485                 269  61.87   
3  RG Pollock        23    41         4  2256                 274  60.97   
4  GA Headley        22    40         4  2190                 270  60.83   

   balls_faced  batting_strike_rate  100  50  0   4s  6s  start_date  \
0         9800                58.60   29  13  7  626   6        1928   
1         1287                91.76    4   7  1  141  23        2022   
2         2667                55.68    5   4  2  186   5        2015   
3         1707                54.48    7  11  1  246  11        1963   
4          416                56.00   10   5  2  104   1        1930   

   end_date country  career_length  
0      1948     AUS             20  
1      2023     ENG              1  
2      2016     AUS              1  
3      1970      SA              7  
4      1954      WI             24

## Answer questions

#### Q1. What is the average `career_length`? 

In [68]:
avg_car_len = round(df.career_length.mean(), 2)
print(f"Average career length is  {avg_car_len}  years")


Average career length is  13.19  years


#### Q2. What is the average `batting_strike_rate` for players who played at least 10 year?

In [71]:
avg_bat_strike = round(df[df['career_length'] >= 10].batting_strike_rate.mean(), 2)
print(f"Average batting_strike_rate is  {avg_bat_strike}")

Average batting_strike_rate is  46.81


#### Q3. What is the number of cricketers who player before 1980?

In [77]:
players_before_1980 = df[df['start_date'] < 1980].player.count()
print(f"There are {players_before_1980} players who played before 1980")

There are 32 players who played before 1980


#### Q4. What is the Max `highest_inns_score` by Country?

In [82]:
df.groupby('country')['highest_inns_score'].max().to_frame('Max_highest_inns_score').reset_index().sort_values('Max_highest_inns_score', ascending = False)

country  Max_highest_inns_score
5    ICC/WI                     400
0       AUS                     380
10       SL                     374
11       WI                     365
1       ENG                     364
3   ICC/PAK                     329
2   ICC/IND                     319
8       PAK                     313
9        SA                     278
4    ICC/SA                     277
6       IND                     254
7        NZ                     251
12      ZIM                     232

#### Q5. What is the average `100`, `50`, `0` by Country?

In [87]:
df.groupby('country')[['100', '50', '0']].mean().sort_values(by = '100', ascending = False)

100         50          0
country                                 
ICC/SA   36.000000  48.000000  13.500000
ICC/WI   34.000000  48.000000  17.000000
ICC/IND  29.500000  47.500000  12.000000
IND      29.500000  36.500000  10.750000
SL       28.666667  44.000000  12.333333
ICC/PAK  25.000000  46.000000  15.000000
PAK      21.250000  28.000000   9.500000
AUS      20.625000  28.312500   8.500000
NZ       18.000000  21.000000   5.500000
WI       16.625000  25.625000   7.250000
ENG      12.307692  20.769231   4.307692
ZIM      12.000000  27.000000   5.000000
SA        9.800000  20.200000   3.400000